pkg/clusteragent/admission/mutate/cws_instrumentation.go

```diff
@@ -0,0 +1,314 @@
+				injected = true
+				return nil
+			}
+		}
+	}
+
+	// override the command with the call to cws-instrumentation
+	exec.Command = append([]string{
+		filepath.Join(cwsMountPath, "cws-instrumentation"),
+		"inject",
+		"--session-type",
+		"k8s",
+		"--data",
+		string(userSessionCtx),
+		"--",
+	}, exec.Command...)
+
+	log.Debugf("Pod exec request to %s is now instrumented for CWS", podString(pod))
+	injected = true
+
+	return nil
+}
+
+// InjectCWSPodInstrumentation injects CWS pod instrumentation
+func InjectCWSPodInstrumentation(rawPod []byte, _ string, ns string, _ *authenticationv1.UserInfo, dc dynamic.Interface, _ kubernetes.Interface) ([]byte, error) {
+	return mutate(rawPod, ns, injectCWSPodInstrumentation, dc)
+}
+
+func injectCWSPodInstrumentation(pod *corev1.Pod, ns string, _ dynamic.Interface) error {
+	var injected bool
+	defer func() {
+		metrics.MutationAttempts.Inc(metrics.CWSPodInstrumentation, strconv.FormatBool(injected), "")
+	}()
+
+	if pod == nil {
+		metrics.MutationErrors.Inc(metrics.CWSPodInstrumentation, "nil pod", "")
+		return fmt.Errorf("cannot inject CWS instrumentation into nil pod")
+	}
+
+	// sanity check: make sure the provided CWS Injector image name isn't empty
+	if len(config.Datadog.GetString("admission_controller.cws_instrumentation.image_name")) == 0 {
+		metrics.MutationErrors.Inc(metrics.CWSPodInstrumentation, "missing image name", "")
+		// don't touch the pod and leave now
+		return fmt.Errorf("empty CWS Injector image name")
+	}
+
+	// compute init container resources
+	resources, err := initCWSInitContainerResources()
+	if err != nil {
+		metrics.MutationErrors.Inc(metrics.CWSPodInstrumentation, "invalid resources configuration", "")
+		return err
+	}
+
+	// is the namespace targeted by the instrumentation ?
+	if !isNsTargetedByCWSInstrumentation(ns) {
+		return nil
+	}
+
+	annotations := pod.GetAnnotations()
+
+	// check if the pod has already been instrumented
+	if isPodCWSInstrumentationReady(annotations) {
+		injected = true
+		// nothing to do, return
+		return nil
+	}
+
+	// create a new volume that will be used to share cws-instrumentation across the containers of this pod
+	injectCWSVolume(pod)
+
+	// bind mount the volume to all the containers of the pod
+	for i := range pod.Spec.Containers {
+		injectCWSVolumeMount(&pod.Spec.Containers[i])
+	}
+
+	// add init container to copy cws-instrumentation in the cws volume
+	injectCWSInitContainer(pod, resources)
+
+	// add label to indicate that the pod has been instrumented
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+	annotations[cwsInstrumentationPodAnotationStatus] = cwsInstrumentationPodAnotationReady
+	pod.Annotations = annotations
+	injected = true
+	log.Debugf("Pod %s is now instrumented for CWS", podString(pod))
+	return nil
+}
+
+func injectCWSVolume(pod *corev1.Pod) {
+	// make sure that the cws volume doesn't already exists
+	for _, vol := range pod.Spec.Volumes {
+		if vol.Name == cwsVolumeName {
```

Comment:
The volume exists but does it have the expected configuration? Note that in any case the output will be diffed, so setting the same thing again won't hurt.

Reply:
Got it. Updated 👍🏻